### ISSUE TITLE
Tracking issue for supporting macOS on Apple Silicon (a.k.a arm64, M1, M2, aarch64)

### ISSUE THREAD

**Issue by shepmaster** [rust-lang/rust#73908]
On 2020-06-22, [Apple announced][announce] that future computers will utilize Arm processors instead of x86_64 processors. This is the tracking issue for supporting that platform, commonly referred to as **Apple Silicon**.

> :warning: **Don't rely on this thread to decide whether to purchase Apple Silicon hardware.** :warning:
> The Rust project does not make any guarantee on the level of support the compiler has.

[announce]: https://www.apple.com/newsroom/2020/06/apple-announces-mac-transition-to-apple-silicon/

### Current status summary (2022-07-05)

- There is no timeline for any level of support.

- `aarch64-apple-darwin` is a tier 2 target and is available for download via rustup. This compiler can be used to cross-compile from `x86_64-apple-darwin` as well as running natively on Arm.

- Compiler tests are **not** being run at this point, as CI has no way of executing them. This is a blocker for Tier 1 support.

- The x86_64 version of `rustc` works under [Rosetta][], producing x86_64 binaries that themselves run under Rosetta. 

[Rosetta]: https://developer.apple.com/documentation/apple_silicon/about_the_rosetta_translation_environment

### About tracking issues

Tracking issues are used to record the overall progress of implementation. They are also used as hubs connecting to other relevant issues, e.g., bugs or open design questions.

A tracking issue is *not* meant for large scale discussion, questions, or bug reports about a feature. Instead, join the discussion and/or contribute in the [**#t-compiler/arm** Zulip stream][stream] or open a new issue for bug reports. You are encouraged to comment here to link to related issues.

This tracking issue will be closely moderated in an attempt to maintain a high signal-to-noise ratio. 

[stream]: https://rust-lang.zulipchat.com/#narrow/stream/242906-t-compiler.2Farm

### Steps

This is not yet intended to be an exhaustive or even accurate list of steps.

- [ ] ~~Find out if there is anyone inside the compiler team with the time to drive the effort forward. This will allow us to efficiently allocate the [Developer Transition Kits][dtk] (DTK), if we end up receiving any~~. There are no members of the compiler team actively working on this
- [x] Reach out to our contacts at Apple to see if DTKs are available, and to establish a line of communication if there are problems down the line. **Please coordinate with the leads of the compiler and infra teams on Zulip on this**.
- [x] Achieve Tier 3 support:
  - ~~Wait for proper support to be upstreamed into LLVM, and update the Rust compiler to pull the latest version of LLVM~~.
    - Our current LLVM supports targeting the appropriate hardware well enough.
  - Update the [libc crate][libc]'s conditional compilation to account for the new target.
    - [x] https://github.com/rust-lang/libc/pull/1817 (in libc 0.2.73)
  - Update the [cc crate][cc] to account for the new target.
    - [x] https://github.com/alexcrichton/cc-rs/pull/527 (in cc 1.0.58) 
  - Submit an initial target specification for `aarch64-apple-darwin`.
    - [x] https://github.com/rust-lang/rust/pull/74541
  - Get the target specification into the bootstrap compiler.
    - [x] https://github.com/rust-lang/rust/pull/74541
    - [x] https://github.com/rust-lang/rust/pull/75261
- [x] Achieve Tier 2 support:
  - [x] Get Xcode 12 on our CI
  - Update the [openssl-src crate][openssl-src] to account for the new target.
    - [x] https://github.com/alexcrichton/openssl-src-rs/pull/74
  - Setup a cross-compiling builder to reach Tier 2
    - [x] Add to CI (https://github.com/rust-lang/rust/pull/75991)
    - [x] Build Cargo (https://github.com/rust-lang/rust/pull/77239)
    - [x] Distribute binaries (https://github.com/rust-lang/rust/pull/77895)
- [ ] Achieve Tier 1 support:
  - [ ] Entire test suite passes
    - [x] #78092
    - [x] #78881
  - [ ] Increase the quality and reliability of the target
  - [ ] Hook up target hardware to our CI, and start running the test suite as part of our standard QA process.
  - [ ] Propose the promotion to Tier 1.
- Miscellaneous
  - Rustup
    - [x] Add support to rustup (https://github.com/rust-lang/rustup/pull/2521)
    - [x] Release it (https://github.com/rust-lang/rustup/issues/2413)

[dtk]: https://developer.apple.com/programs/universal/
[libc]: https://github.com/rust-lang/libc
[cc]: https://github.com/alexcrichton/cc-rs
[openssl-src]: https://github.com/alexcrichton/openssl-src-rs

### Resolved Questions

- The infra team does **not** want to maintain our own hardware / VMs. This means solutions like MacStadium, AWS, or "putting our own hardware under someone's desk" are non-starters.

### Unresolved Questions

This is not yet intended to be an exhaustive or even accurate list of questions.

- [x] When will our CI providers have access to native hardware?
  - [x] https://github.com/actions/virtual-environments/issues/2187
  - [x] https://github.com/github/roadmap/issues/528
- [ ] Should we use an alternative / additional CI provider to GitHub?
  - [ ] [Cirrus CI](https://cirrus-ci.org/guide/macOS)
- [ ] What is an accurate target name for the new platform?
  - [ ] What is the accurate name of the chip ("silicon", "aarch64", "arm64", etc.)?
- [ ] How can we leave room to support Apple's `arm64e`?

**Comment by hasanihunter** [comment]
For those devs that have a DTK in their possession, is there any way they can help?

**Comment by shepmaster** [comment]
>> is there any way they can help

> join the discussion in the [**#t-compiler/arm** Zulip stream](https://rust-lang.zulipchat.com/#narrow/stream/242906-t-compiler.2Farm)

**Comment by rudyrichter** [comment]
@shepmaster is there any help needed getting to the point where one can cross-compile targeting the new architecture from an existing one (macOS x86_64 cross compiling for macOS arm64) ?

**Comment by shepmaster** [comment]
@rudyrichter no, we have that capability already — see the "Current status summary" section in the top post. Please feel free to join Zulip if you have more questions.

**Comment by shepmaster** [comment]
# Status update

Thanks for your patience! Progress has continued at a suitable pace. If you haven't been refreshing the page manually every few hours, you can see a summary of the work to date in the "Current status summary" section in the top post. The most recent update is that we have opened [a pull request](https://github.com/rust-lang/rust/pull/74541) to add an appropriate target to the compiler codebase. Next steps include getting that merged into nightly and then present in the bootstrapping compiler. Once that's available, we will be able to look into getting builds as part of CI.

---

As a reminder:

> A tracking issue is *not* meant for large scale discussion, questions, or bug reports about a feature. Instead, join the discussion and/or contribute in the [**#t-compiler/arm** Zulip stream][stream] or open a new issue for bug reports. You are encouraged to comment here to link to related issues.

[stream]: https://rust-lang.zulipchat.com/#narrow/stream/242906-t-compiler.2Farm

**Comment by luxxxxy** [comment]
> How can we leave room to support Apple's arm64e?

I am currently adding PAC/arm64e to a [fork using Apple's LLVM](https://github.com/Crabapple-iOS/rust/tree/arm64e-apple-ios). Hopefully some of my work can be reused whenever PAC/arm64e support is upstreamed into LLVM.

**Comment by glandium** [comment]
Not sure if I should open a separate issue, but running the x86-64 rustc on a DTK (via Rosetta 2) yields `<jemalloc>: Unsupported system page size` messages (but things still work), so it's likely that the arm64 port would need some adjustment in that area. (Not sure if the x86-64 rustc should be adjusted to to work without those messages under Rosetta 2)

(The issue being that the page size is 16KB instead of 4KB)

**Comment by glandium** [comment]
For the record, I've built the rust compiler on the DTK with https://github.com/rust-lang/rust/pull/75500, a tweak to stage0.txt to pick a recent nightly, and a config.toml that sets the following:
```
build = "x86_64-apple-darwin"
host = ["aarch64-apple-darwin"]
target = ["aarch64-apple-darwin"]
```
and was subsequently able to build all the rust code in Firefox *but* I had to make changes to three crates that were passing the wrong --target to clang because the rust target differs from the clang target (rust uses aarch64-apple-darwin, clang uses arm64-apple-darwin). This is very unfortunate, and it's probably going to cause problems in other crates.

**Comment by pnkfelix** [comment]
I'm throwing a link to https://github.com/rust-lang/rust/issues/74820 on here because I suspect we'll want to better understand *that* bug before we can stabilize a Tier 1 ARM target.

**Comment by aspenluxxxy** [comment]
> I'm throwing a link to #74820 on here because I suspect we'll want to better understand _that_ bug before we can stabilize a Tier 1 ARM target.

Does _that_ bug affect any 64-bit ARM targets? Or just ARMv7?

**Comment by glandium** [comment]
> Get Xcode 12 on our CI, and setup a cross-compiling builder to reach Tier 2.

It's already there, both on github actions and azure pipelines, per https://github.com/actions/virtual-environments/blob/main/images/macos/macos-10.15-Readme.md and https://github.com/microsoft/azure-pipelines-image-generation/blob/master/images/macos/macos-10.15-Readme.md (which look like to be the same document, btw)

**Comment by shepmaster** [comment]
# Status update

Thanks for your patience! Progress has continued at a slow pace. You can see a summary of the work to date in the "Current status summary" section in the top post. The most recent update is that we have opened [a **draft** pull request](https://github.com/rust-lang/rust/pull/75991) to add a cross-compiler builder as part of CI. Once that is reviewed and hopefully merged, then artifacts should be available to download. It is important to note that the compiler tests **are not** being run at this point, as CI has no way of executing them.

---

As a reminder:

> A tracking issue is *not* meant for large scale discussion, questions, or bug reports about a feature. Instead, join the discussion and/or contribute in the [**#t-compiler/arm** Zulip stream][stream] or open a new issue for bug reports. You are encouraged to comment here to link to related issues.

[stream]: https://rust-lang.zulipchat.com/#narrow/stream/242906-t-compiler.2Farm

**Comment by shepmaster** [comment]
# Status update

Progress has continued at a measured pace. You can see a summary of the work to date in the "Current status summary" section in the top post. The most recent update is that aarch64-apple-darwin is **now a Tier 2 platform**. It is important to note that the compiler tests **are not** being run at this point, as CI has no way of executing them.

If you'd like to try using the new compiler, I've created some [basic instructions][] on how to configure your project to cross-compile to aarch64-apple-darwin. If you have access to an Apple DTK, there are also instructions on how to download and install the native toolchain.

[basic instructions]: https://github.com/shepmaster/rust/blob/silicon/silicon/README.md

---

As a reminder:

> A tracking issue is *not* meant for large scale discussion, questions, or bug reports about a feature. Instead, join the discussion and/or contribute in the [**#t-compiler/arm** Zulip stream][stream] or open a new issue for bug reports. You are encouraged to comment here to link to related issues.

[stream]: https://rust-lang.zulipchat.com/#narrow/stream/242906-t-compiler.2Farm

**Comment by agentsim** [comment]
The instructions to install rustup under the section "Compiling on native aarch64 hardware" no longer seem to work. Is there an updated command?

**Comment by glandium** [comment]
> The instructions to install rustup under the section "Compiling on native aarch64 hardware" no longer seem to work.

What kind of error are you getting?

**Comment by agentsim** [comment]
I'm seeing this:
```
me@my-Mac ~ % arch --x86_64 \
sh <(curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs) \
--default-host aarch64-apple-darwin \
--default-toolchain nightly
arch: posix_spawnp: sh: Bad CPU type in executable
me@my-Mac ~ % curl: (23) Failed writing body (0 != 710)
```
I just noticed an update to BigSur 11.0.1, I'll try updating. But these instructions seem to predate that release.

**Comment by glandium** [comment]
That command still works for me on beta 10.

**Comment by HenkPoley** [comment]
It seems your `sh` is not a fat binary with x86_64. Did you build one yourself ? Maybe write the full path to Apple's `sh`?

**Comment by agentsim** [comment]
Looks like this is a problem with my DTK:
```
me@my-Mac ~ % lipo -info /bin/sh
Architectures in the fat file: /bin/sh are: x86_64 arm64e 
me@my-Mac ~ % arch -arch x86_64 /bin/sh
arch: posix_spawnp: /bin/sh: Bad CPU type in executable
```

**Comment by shepmaster** [comment]
# Status update

Progress in the compiler proper has slowed, but progress in the rest of the Rust ecosystem is increasing as there is now commercially-available hardware. You can see a summary of the work to date in the "Current status summary" section in the top post. The most recent update is that [rustup is now available for aarch64-apple-darwin][available]. This means that the usual [Rust installation instructions][install] should now work as-is. Note that no native stable build has been produced yet, so you will need to use the beta or nightly channel until Rust 1.49 is released.

If you'd like to try using the new compiler, I've created some [basic instructions][] on how to configure your project to cross-compile to aarch64-apple-darwin. If you have access to a machine with an Apple M1 chip, there are also instructions on how to download and install the native toolchain.

[available]: https://blog.rust-lang.org/2020/11/27/Rustup-1.23.0.html
[install]: https://rustup.rs
[basic instructions]: https://github.com/shepmaster/rust/blob/silicon/silicon/README.md

---

As a reminder:

> A tracking issue is *not* meant for large scale discussion, questions, or bug reports about a feature. Instead, join the discussion and/or contribute in the [**#t-compiler/arm** Zulip stream][stream] or open a new issue for bug reports. You are encouraged to comment here to link to related issues.

[stream]: https://rust-lang.zulipchat.com/#narrow/stream/242906-t-compiler.2Farm

**Comment by glandium** [comment]
> I've created some basic instructions

As issues are not opened on that repo, I'll leave a note here: No need for Xcode beta now, the latest release is fine. But Xcode is actually not needed either, only the command line tools. In any case, if someone has a setup to build rust on a x86_64 mac, and is up-to-date, they don't need to do anything about either Xcode or CLT.

The `RUSTC=$(rustup which rustc) $(rustup which cargo) build` trick is only necessary when rustup itself is x86_64, which it is not anymore. It would be better to put instructions to check whether rustup is native or not, and how to update if it isn't (sadly, `rustup self update` won't fix it ; it would if rustup was a universal binary)

**Comment by shepmaster** [comment]
> As issues are not opened on that repo, I'll leave a note here

Thanks! PRs should be open there, even if issues aren't. You can also ping me in the [**#t-compiler/arm** Zulip stream](https://rust-lang.zulipchat.com/#narrow/stream/242906-t-compiler.2Farm).

**Comment by chadbrewbaker** [comment]
This seemed to work:
```bash
curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- --default-toolchain nightly
```

**Comment by shepmaster** [comment]
> [Note that no native stable build has been produced yet, so you will need to use the beta or nightly channel until Rust 1.49 is released.](https://github.com/rust-lang/rust/issues/73908#issuecomment-734918768)

**Comment by frewsxcv** [comment]
I opened a pull request to enable sanitizers for Apple Silicon https://github.com/rust-lang/rust/pull/79883

**Comment by monokrome** [comment]
fwiw, Apple silicon macs report themselves as **arm64**:

```
# uname -a
Darwin eridani.local 20.2.0 Darwin Kernel Version 20.2.0: Wed Dec  2 20:40:21 PST 2020; root:xnu-7195.60.75~1/RELEASE_ARM64_T8101 arm64
```

**Comment by aspenluxxxy** [comment]
@monokrome 

1. yeah, aarch64 is usually only used by compilers and such.
2. also, run `file` on a system app like safari, it'll be `arm64e`.

**Comment by mlindner** [comment]
What's blocking this from getting closed and us getting to tier1? The mentioned issues are both closed despite them not being marked as such. Is the issue not enough available M1 machines for the CI? Is monetary support needed to buy the machines? This hasn't received any updates in a while and people keep referencing the issue to close issues in other repositories.

**Comment by bjorn3** [comment]
See the issue description. I think the biggest thing left to do is:

> Hook up target hardware to our CI, and start running the test suite as part of our standard QA process.

**Comment by shepmaster** [comment]
# Status update

There has been no progress or change to our support, so there is nothing to report. If something changes, we will report it.

As the [primary comment][] has stated for a while, we are waiting on CI infrastructure to exist. We are aware of no CI providers offering the ability to execute aarch64-apple-darwin binaries. The Rust infrastructure team is not particularly interested in becoming our own CI provider, either. If you are the owner or operator of a CI platform that supports native execution, please feel free to reach out to us (instructions below).

[primary comment]: https://github.com/rust-lang/rust/issues/73908#issue-648613557

---

As a reminder:

> A tracking issue is *not* meant for large scale discussion, questions, or bug reports about a feature. Instead, join the discussion and/or contribute in the [**#t-compiler/arm** Zulip stream][stream] or open a new issue for bug reports. You are encouraged to comment here to link to related issues.

[stream]: https://rust-lang.zulipchat.com/#narrow/stream/242906-t-compiler.2Farm

**Comment by aspenluxxxy** [comment]
EC2 has M1s, but yeah, I don't think y'all would want to be the ones to manage them though.

**Comment by tarcieri** [comment]
They do? The [Amazon EC2 Mac Instances](https://aws.amazon.com/ec2/instance-types/mac/) page only lists i7.

**Comment by aspenluxxxy** [comment]
@tarcieri wait, I misremembered, M1 EC2 instances are planned for sometime this year. my bad!!

**Comment by richiksc** [comment]
[MacStadium has cloud M1s.](https://www.macstadium.com/mac-mini)

**Comment by iMonZ** [comment]
Platform name: macOS
Chip: arm64, arm64e or UB (Universal Binary)

**Comment by pjuangph** [comment]
I was able to resolve the problem by doing this https://stackoverflow.com/a/69071041/1599606 

I definitely agree with the OP that we need better ARM64 support moving forward.

**Comment by ShadowJonathan** [comment]
If this wasn't already posted here, something relevant; https://github.com/actions/virtual-environments/issues/2187

**Comment by sdavids** [comment]
Status? 

Since the last update several CI providers have Apple M1 VM offerings.

**Comment by shepmaster** [comment]
@sdavids  please educate me on which *CI providers* allow executing Apple Silicon binaries. As the previous comment emphasizes, we are not interested in bare hardware providers (e.g. AWS EC2 / Scaleway / MacStadium). There are multiple (off topic and thus hidden) comments about those providers.

**Comment by tarcieri** [comment]
I believe Apple's own [Xcode Cloud (Beta)](https://developer.apple.com/xcode-cloud/beta/) service might offer this? They describe it as:

> Xcode Cloud is a new continuous integration and delivery service built into Xcode 13, and designed expressly for Apple developers.

**Comment by shepmaster** [comment]
Speaking as myself, **not for the infra team**, I'd be highly uncomfortable using Xcode Cloud based on my current understanding of it. As you mention, it's beta. I also expect that it's highly focused on providing a good experience for people with more of a buy-in to the Xcode ecosystem. From [Requirements for Using Xcode Cloud](https://developer.apple.com/documentation/xcode/requirements-for-using-xcode-cloud):

> To use Xcode Cloud, be sure to meet the following project and workspace requirements:
> - You use an Xcode project or workspace.
> [...]
> - You use Xcode’s new build system. [...]
> - Your dependencies and additional third-party tools are available to Xcode Cloud. [...]

From [Making Dependencies Available to Xcode Cloud](https://developer.apple.com/documentation/xcode/making-dependencies-available-to-xcode-cloud), it's _maybe_ possible to hack up a bunch of shell scripts to call back out to `x.py` and friends, but it's certainly not the intended integration path. If someone has access to Xcode Cloud, wants to investigate building Rust there, and is OK with potentially any work not being upstreamed, please join the discussion in the [**#t-compiler/arm** Zulip stream][stream].

[stream]: https://rust-lang.zulipchat.com/#narrow/stream/242906-t-compiler.2Farm

**Comment by sdavids** [comment]
https://github.com/github/roadmap/issues/240

**Comment by taiki-e** [comment]
For CI:

- AWS now has macOS M1 instance (preview): https://aws.amazon.com/en/blogs/aws/use-amazon-ec2-m1-mac-instances-to-build-test-macos-ios-ipados-tvos-and-watchos-apps/
- Cirrus CI's macOS VM is now M1: https://cirrus-ci.org/guide/macOS

(FWIW, we are using the later in [crossbeam's CI](https://github.com/crossbeam-rs/crossbeam/blob/master/.cirrus.yml#L22-L39))

**Comment by pnkfelix** [comment]
Discussed in T-compiler [backlog bonanza](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bsteering.20meeting.5D.202022-09-09.20compiler-team.23484.20backlog.20b'za/near/297991504)

Seems like Ci support might be the main known thing blocking pushing this up to Tier 1 support. Which is awesome! (But still, that is an implementation step that remains incomplete.)

@rustbot label: S-tracking-impl-incomplete

**Comment by tarcieri** [comment]
GitHub recently added support for [self-hosted M1 runners](https://github.blog/changelog/2022-08-09-github-actions-self-hosted-runners-now-support-apple-m1-hardware/) 

GitHub-hosted M1 runners are on their roadmap: https://github.com/github/roadmap/issues/528

**Comment by controversial** [comment]
Note the GitHub roadmap item for macOS ARM runners (https://github.com/github/roadmap/issues/528) was moved from “future” to “Q3 2023”

**Comment by christian-steinmeyer** [comment]
@controversial The roadmap issue explicitly contains "M1" in the title, but doesn't mention "M2". Do you have any idea, whether it'll cover both chips?

**Comment by bjorn3** [comment]
M2 is backwards compatible with M1, so any code running on an M1 will also run on an M2.

**Comment by aviramha** [comment]
Is there any discussion about supporting arm64e as target output?

**Comment by bjorn3** [comment]
Isn't arm64e still restricted to apple signed binaries unless you use the `-arm64e_preview_abi` boot argument as the ABI isn't stabilized yet?

**Comment by aviramha** [comment]
> Isn't arm64e still restricted to apple signed binaries unless you use the `-arm64e_preview_abi` boot argument as the ABI isn't stabilized yet?

It is, but you can still write `dylib`s to integrate with existing binaries and then you need to be arm64e.

**Comment by hdost** [comment]
Is this something Apple would be willing to sponsor?
> Hook up target hardware to our CI, and start running the test suite as part of our standard QA process.

Not sure if they'd be too inclined given they mostly push Swift.

**Comment by aviramha** [comment]
> Is this something Apple would be willing to sponsor?
> > Hook up target hardware to our CI, and start running the test suite as part of our standard QA process.
> 
> Not sure if they'd be too inclined given they mostly push Swift. 

I think GitHub might have alpha/beta runner of M1. I'll shoot a tweet maybe someone there help us

**Comment by glandium** [comment]
https://github.com/github/roadmap/issues/528 says Q4 2023, but it was Q3 before that, so who knows how much else it's going to be delayed.

**Comment by aviramha** [comment]
> https://github.com/github/roadmap/issues/528 says Q4 2023, but it was Q3 before that, so who knows how much else it's going to be delayed.

I believe it's already available in some sort of early access and if we push for it we can get rustlang to be an early adopter. Maybe I'm too optimistic but it'd be the best solution (as GitHub is already being used)

**Comment by aviramha** [comment]
Small progress: COO of GitHub responded to my ask on Twitter and said he's checking if he can help!

**Comment by pietroalbini** [comment]
@aviramha please refrain from contacting GitHub (or any other company) on behalf of the Rust project.

The Rust Infrastructure Team has regular calls with GitHub, and M1 runners are a recurrent topic of those calls. People outside the Infrastructure Team setting up backchannels does not help getting M1 runners faster, it just makes the situation confusing and hurts the relationship Rust has built with GitHub.

Please stop the conversation you're having and redirect it to infra@rust-lang.org (if that's via email) or to [@pietroalbini](https://twitter.com/pietroalbini) on Twitter (if it's on Twitter, so that I can route it to the most appropriate place). Thank you.

Pietro.
Rust Infrastructure Team Lead

**Comment by aviramha** [comment]
> @aviramha please refrain from contacting GitHub (or any other company) on behalf of the Rust project.
> 
> 
> 
> The Rust Infrastructure Team has regular calls with GitHub, and M1 runners are a recurrent topic of those calls. People outside the Infrastructure Team setting up backchannels does not help getting M1 runners faster, it just makes the situation confusing and hurts the relationship Rust has built with GitHub.
> 
> 
> 
> Please stop the conversation you're having and redirect it to infra@rust-lang.org (if that's via email) or to [@pietroalbini](https://twitter.com/pietroalbini) on Twitter (if it's on Twitter, so that I can route it to the most appropriate place). Thank you.
> 
> 
> 
> Pietro.
> 
> Rust Infrastructure Team Lead

Hi 👋🏼 
Sorry about this. I didn't contact in name of the organization nor the foundation. I just asked if someone can help as the latest replies asked about having someone sponsor it. I'll send him your handle.
On a personal note I think the tone of your message is quite discouraging while I understand my attempt might have been annoying it was out of desire to help and I didn't have any bad intentions.

**Comment by lunabunn** [comment]
> The Rust Infrastructure Team has regular calls with GitHub, and M1 runners are a recurrent topic of those calls. People outside the Infrastructure Team setting up backchannels does not help getting M1 runners faster, it just makes the situation confusing and hurts the relationship Rust has built with GitHub.

Unless I am misunderstanding, there has not been a single message from you or any other member of the infrastructure team (except for, of course, shepmaster) in this entire issue thread. While I fully understand that internal discussion happens on Zulip and that fractured efforts are oftentimes worse than no effort, it is difficult to comprehend your mention of "backchannels" when said "backchannel" was communicated here and the "recurrent ... relationship Rust has built with GitHub" has not. The most status update even explicitly points out the lack of CI infrastructure as a blocker.

Please remember that we (the community -- though I myself am not a direct contributor to rust-lang/rust) use and contribute to Rust and its ecosystem with good intentions and depend on communication from the Rust teams to coordinate our efforts ❤️

**Comment by darthdeus** [comment]
> @aviramha please refrain from contacting GitHub (or any other company) on behalf of the Rust project.
> 
> The Rust Infrastructure Team has regular calls with GitHub, and M1 runners are a recurrent topic of those calls. People outside the Infrastructure Team setting up backchannels does not help getting M1 runners faster, it just makes the situation confusing and hurts the relationship Rust has built with GitHub.
> 
> Please stop the conversation you're having and redirect it to [infra@rust-lang.org](mailto:infra@rust-lang.org) (if that's via email) or to [@pietroalbini](https://twitter.com/pietroalbini) on Twitter (if it's on Twitter, so that I can route it to the most appropriate place). Thank you.
> 
> Pietro. Rust Infrastructure Team Lead

I'm not sure what you're hoping to achieve with comments like this other than alienate possible contributors. People in the community are having problems, those people are trying to resolve problems on their own because there is no other apparent progress. If the infrastructure team has been working on resolving the problem behind closed doors and nobody knows that it's being worked on, maybe it shouldn't be surprising that the community is trying to resolve the problem on their own.

What actually hurts Rust the language is the constant drama from lacking communication on the leadership side. Comments like this actively discourage any active contributors and add to the already pretty big fire around the Rust project.

Many important people have already left Rust on not so good terms based on how they've been treated by the leadership. Comments like this are only going to make more people leave.

**Comment by infiniteregrets** [comment]
> @aviramha please refrain from contacting GitHub (or any other company) on behalf of the Rust project.

I don't understand this negative enforcement of efforts made by the community, the intentions here, clearly were to help

**Comment by mlindner** [comment]
@pietroalbini Seeing as you appear to be in the know, can you please explain the current status? The top comment hasn't been updated in about 9 months. This is a relatively common topic of conversation so if you know something, please mention it. "We are having conversations" doesn't really say anything. (In general this whole Zulip this Zulip that is a huge negative of the Rust project. Stick to GitHub.)

**Comment by shepmaster** [comment]
> The top comment hasn't been updated in about 9 months

That is because there **is no update** to the status.

I have been using Apple hardware since I was 4 years old. I worked with people with access to the developer kits to get Rust compiling on / for Apple Silicon as soon as possible. I'd believe I want this issue closed more than the vast majority of people subscribed.

As soon as something worth posting is available and I'm awake / available / near a computer, I will post it!

> "We are having conversations" doesn't really say anything

I am not going to post a status update every week / two weeks / month that says "no progress". That will drive people batty and provides no information. GitHub has been a gracious sponsor and the Rust project uses a lot of GitHub functionality; we are lucky to have a few points of contact that we communicate with on an ongoing basis (think email / virtual meetings — I actually choose to not attend those meetings so I'm not 100% sure).

Anything that GitHub chooses to say will be on their public issues, which is why we've linked them above. You are highly encouraged to subscribe there and you will know the same things we do as soon as we do.

> In general this whole Zulip this Zulip that is a huge negative of the Rust project. Stick to GitHub.

The [Rust Zulip](https://rust-lang.zulipchat.com/#narrow/stream/242791-t-infra) is open to read. Posting does require signing up, but you can sign up using your existing GitHub account. Sometimes communication is longer form and sometimes it's shorter form; that's just how we humans work. Zullip represents a different speed of communication than GitHub.

Also, sometimes communications **needs** to be private. This is especially true for a team like the infrastructure team, which has to deal with accounts or passwords or signing keys, etc. Just as you don't post all of your usernames and passwords to a publicly available GitHub repository, neither do we. I get that this can be frustrating — I wish I could know every sordid detail of GitHub's M1 implementation, but I don't get to and that's just something I have to deal with.

> the infrastructure team doesn't want to maintain actual physical infrastructure

The team is small and has a finite amount of time and experience so we have to prioritize what we work on. Creating a production-hardened macOS CI system that's open to arbitrary attackers from the internet is not trivial and is not cheap.

> the meaning of the word infrastructure implies something physical

One thing that I have had to deal with is that words are created or change and take on new meanings (no cap, fam). Other than my own machines, I think I last helped rack a physical server about twenty years ago. I can't imagine any of our volunteers hopping into their car and driving down to the colo to swap out a hard drive after their pager woke them up at 3 in the morning so that every Rust build in the world can continue.

> expecting GitHub to provide every piece of infrastructure

I don't disagree: having our eggs in multiple baskets *would* be nice. However, this comes back to the time tradeoff. The Rust build is nontrivial and there's no easy button to press to make it distributed across multiple CI providers. We've actually had that in the past and breathed a sigh of relief when we could consolidate back to one provider.

> The Linux project seems to manage it somehow

Certainly, this isn't something impossible that breaks the laws of physics, it's just *hard*. I have a feeling that Linux is a bigger project and better resourced.

---

If you are thinking to yourself "I've got a good amount of time to contribute and have the skills to maintain / improve Rust's infrastructure", then by all means come on [over to Zulip](https://rust-lang.zulipchat.com/#narrow/stream/242791-t-infra) and *talk with us*. 

If you are a company that provides Apple Silicon macOS CI and want to donate your services, talk to us and/or the Rust Foundation and we'll see if we can make anything happen.

Ultimately, please remember that the infra team is composed of **humans** with lives and thoughts and feelings. We aren't hiding things from you, we aren't maliciously degrading Rust CI, etc. I'm on vacation this week and instead of spending time with my daughter this morning, I'm replying to this issue in the hopes of reducing tensions so that there's not a bigger fire later on. I hope that I don't regret this morning's time choices.

**Comment by mlindner** [comment]
Thank you for the long and detailed response. Apologies from taking you away from your vacation. Please have and have a nice day.

**Comment by shepmaster** [comment]
GitHub [recently announced][announcement] the public beta availability of Apple Silicon builders for GitHub Actions. We are as excited as you are about this news, and we plan on trying them out as soon as possible! After we have had a chance to do so, we'll be sure to report back here. Thanks for hanging in there for as long as you have, and we hope to have more good news in the near future.

[announcement]: https://github.blog/2023-10-02-introducing-the-new-apple-silicon-powered-m1-macos-larger-runner-for-github-actions/

**Comment by rami3l** [comment]
Small update: Recently we have been running the full Rustup test suite on `aarch64-apple-darwin` and so far it's doing great: https://github.com/rust-lang/rustup/pull/3681

**Comment by shepmaster** [comment]
I have opened [RFC 3671](https://github.com/rust-lang/rfcs/pull/3671) to promote aarch64-apple-darwin to Tier 1.

**Comment by boozook** [comment]
> When will our CI providers have access to native hardware?

Just have to notice that GitHub CI has support M1 runners - "macos-14".

**Comment by evelynharthbrooke** [comment]
PR #128592 promotes `aarch64-apple-darwin` to Tier 1, as per the newly merged RFC.

**Comment by evelynharthbrooke** [comment]
As of today, ARM64 macOS (Darwin) is an officially recognized tier 1 target. Should this issue be closed now as successful in its main goal of having arm64 macOS be an officially supported Tier 1 target?

**Comment by evelynharthbrooke** [comment]
As of today, ARM64 macOS (Darwin) is an officially recognized tier 1 target. I am proposing that this issue be closed now as the main goal of having arm64 macOS be an officially supported Tier 1 target has been reached.

**Comment by ehuss** [comment]
Thanks @evelynharthbrooke! I'll go ahead and close as completed by #128592.
